### PR TITLE
Ftr: dynamic tag router

### PR DESCRIPTION
**Pull request by watermelo**
**What this PR does**:
#593 

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: cluster/router/tag/router_rule.go [hunk 1]
```diff
@@ -52,18 +52,44 @@ func getRule(rawRule string) (*RouterRule, error) {
...
 	}
 	r.RawRule = rawRule
-	// TODO init tags
+	r.init()
 	return r, nil
 }
 
+func (t *RouterRule) init() {
+	t.addressToTagNames = make(map[string][]string)
+	t.tagNameToAddresses = make(map[string][]string)
```
Comment (AlexStocks) [inline cluster/router/tag/router_rule.go:61]: t.addressToTagNames = make(map[string][]string, 8)
	t.tagNameToAddresses = make(map[string][]string, 8)

File: cluster/router/tag/router_rule.go [hunk 2]
```diff
@@ -52,18 +52,44 @@ func getRule(rawRule string) (*RouterRule, error) {
...
+			t.addressToTagNames[address] = append(t.addressToTagNames[address], tag.Name)
+		}
+		t.tagNameToAddresses[tag.Name] = tag.Addresses
+	}
+}
+
 func (t *RouterRule) getAddresses() []string {
-	// TODO get all tag addresses
-	return nil
+	var result []string
```
Comment (AlexStocks) [inline cluster/router/tag/router_rule.go:71]: var result = make([]string, 0, 8 * len(t.Tags))

File: cluster/router/tag/router_rule.go [hunk 3]
```diff
@@ -52,18 +52,44 @@ func getRule(rawRule string) (*RouterRule, error) {
...
+	for _, tag := range t.Tags {
+		result = append(result, tag.Addresses...)
+	}
+	return result
 }
 
 func (t *RouterRule) getTagNames() []string {
-	// TODO get all tag names
-	return nil
+	var result []string
```
Comment (AlexStocks) [inline cluster/router/tag/router_rule.go:79]: var result = make([]string, 0, len(t.Tags))

File: cluster/router/tag/tag_router.go [hunk 4]
```diff
@@ -172,18 +185,51 @@ func isForceUseTag(url *common.URL, invocation protocol.Invocation) bool {
...
+	}
+	return invokers[:idx]
 }
 
 func checkAddressMatch(addresses []string, host, port string) bool {
 	for _, address := range addresses {
-		// TODO address parse
-		if address == (host + port) {
+		// TODO ip match
+		if address == host+":"+port {
```
Comment (AlexStocks) [inline cluster/router/tag/tag_router.go]: delete "host+":"+port", using net.JoinHostPort(host, port) instead.

File: cluster/router/tag/router_rule.go [hunk 5]
```diff
@@ -22,9 +22,27 @@ import (
...
+ * - name: tag1
+ * addresses: [ip1, ip2]
+ * - name: tag2
+ * addresses: [ip3, ip4]
+ * ...
+ */
 // RouterRule RouterRule config read from config file or config center
 type RouterRule struct {
 	router.BaseRouterRule `yaml:",inline""`
+	Tags                  []Tag
```
Comment (zouyx) [inline cluster/router/tag/router_rule.go:43]:
```suggestion
	tags                  []Tag
```
you mean `tags`?

File: cluster/router/tag/tag.go [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package tag
+
+type Tag struct {
```
Comment (zouyx) [inline cluster/router/tag/tag.go:20]: you mean `tag`?
```suggestion
type tag struct {
```

File: cluster/router/tag/tag_router.go [hunk 7]
```diff
@@ -55,6 +65,16 @@ func (c *tagRouter) isEnabled() bool {
 	return c.enabled
 }
 
+func (c *tagRouter) SetApplication(app string) {
+	c.application = app
+}
+
+func (c *tagRouter) tagRouterRuleCopy() RouterRule {
+	fmt.Println(c.tagRouterRule, "fuck")
```
Comment (zouyx) [inline cluster/router/tag/tag_router.go:73]:
```suggestion
	fmt.Println(c.tagRouterRule, "fuck")
```
i think this feature has made you crazy ：）

File: cluster/router/tag/tag_router.go [hunk 8]
```diff
@@ -63,7 +83,152 @@ func (c *tagRouter) Route(invokers []protocol.Invoker, url *common.URL, invocati
...
+	tagRouterRuleCopy := c.tagRouterRuleCopy()
+	tag, ok := invocation.Attachments()[constant.Tagkey]
+	if !ok {
+		tag = url.GetParam(constant.Tagkey, "")
+	}
+	var (
+		result    []protocol.Invoker
+		addresses []string
+	)
+	if tag != "" {
```
Comment (zouyx) [inline cluster/router/tag/tag_router.go:99]:
```suggestion
	if len(tag)>0  {
```

File: cluster/router/tag/tag_router.go [hunk 9]
```diff
@@ -63,7 +83,152 @@ func (c *tagRouter) Route(invokers []protocol.Invoker, url *common.URL, invocati
...
+func (c *tagRouter) Process(event *config_center.ConfigChangeEvent) {
+	logger.Infof("Notification of tag rule, change type is:[%s] , raw rule is:[%v]", event.ConfigType, event.Value)
+	if remoting.EventTypeDel == event.ConfigType {
+		c.tagRouterRule = nil
+		return
+	} else {
+		content, ok := event.Value.(string)
+		if !ok {
+			msg := fmt.Sprintf("Convert event content fail,raw content:[%s] ", event.Value)
+			logger.Error(msg)
```
Comment (zouyx) [inline cluster/router/tag/tag_router.go:182]: Use logger.Errorf ?

File: cluster/router/tag/tag_router.go [hunk 10]
```diff
@@ -63,7 +83,152 @@ func (c *tagRouter) Route(invokers []protocol.Invoker, url *common.URL, invocati
...
+
+	routerKey := providerApplication + constant.TagRouterRuleSuffix
+	dynamicConfiguration.AddListener(routerKey, c)
+	//get rule
+	rule, err := dynamicConfiguration.GetRule(routerKey, config_center.WithGroup(config_center.DEFAULT_GROUP))
+	if len(rule) == 0 || err != nil {
+		logger.Errorf("Get rule fail, config rule{%s},  error{%v}", rule, err)
+		return
+	}
+	if rule != "" {
```
Comment (zouyx) [inline cluster/router/tag/tag_router.go:226]: len>0

File: cluster/router/tag/tag_router.go [hunk 11]
```diff
@@ -100,3 +265,163 @@ func isForceUseTag(url *common.URL, invocation protocol.Invocation) bool {
...
+			if !filter(invoker) {
+				continue OUTER
+			}
+		}
+		res = append(res, invoker)
+	}
+	return res
+}
+
+// TODO 需要搬到 dubbogo/gost, 可以先 review
```
Comment (zouyx) [inline cluster/router/tag/tag_router.go:285]: Use english?

File: cluster/router/tag/router_rule.go (hunk 5 above)
Comment (watermelo) [reply cluster/router/tag/router_rule.go:43]: tags can't be yaml parse.

Comment (watermelo) [reply cluster/router/tag/router_rule.go:43]: tags is unexport, so can't be yaml parsed

File: cluster/router/tag/tag_router.go [hunk 12]
```diff
@@ -63,7 +81,152 @@ func (c *tagRouter) Route(invokers []protocol.Invoker, url *common.URL, invocati
...
+	// since the rule can be changed by config center, we should copy one to use.
+	tagRouterRuleCopy := c.tagRouterRuleCopy()
+	tag, ok := invocation.Attachments()[constant.Tagkey]
+	if !ok {
+		tag = url.GetParam(constant.Tagkey, "")
+	}
+	var (
+		result    []protocol.Invoker
+		addresses []string
+	)
```
Comment (pantianying) [inline cluster/router/tag/tag_router.go:96]: The definition can be moved to the top.
And can you break this pile of code into several parts,it is hard for me 😂

File: cluster/router/tag/tag_router.go [hunk 13]
```diff
@@ -100,3 +263,163 @@ func isForceUseTag(url *common.URL, invocation protocol.Invocation) bool {
...
+		isIpv4 = false
+	}
+
+	hostAndPort := getPatternHostAndPort(pattern, isIpv4)
+	if hostAndPort[1] != "" && hostAndPort[1] != port {
+		return false
+	}
+
+	pattern = hostAndPort[0]
+	// TODO 常量化
```
Comment (pantianying) [inline cluster/router/tag/tag_router.go:332]: has some easy todo

File: cluster/router/tag/tag_router_test.go [hunk 14]
```diff
@@ -19,32 +19,59 @@ package tag
 
 import (
 	"context"
+	"fmt"
+	"github.com/stretchr/testify/suite"
```
Comment (pantianying) [inline cluster/router/tag/tag_router_test.go:23]: pkg import problem

Comment (pantianying) [inline cluster/router/tag/tag_router_test.go:23]: pkg import problem

File: cluster/router/tag/tag_router.go (hunk 12 above)
Comment (AlexStocks) [reply cluster/router/tag/tag_router.go:96]: agree

File: cluster/router/tag/tag_router.go [hunk 15]
```diff
@@ -100,3 +263,163 @@ func isForceUseTag(url *common.URL, invocation protocol.Invocation) bool {
...
+
+type filter func(protocol.Invoker) bool
+
+func filterInvoker(invokers []protocol.Invoker, filters ...filter) []protocol.Invoker {
+	var res []protocol.Invoker
+OUTER:
+	for _, invoker := range invokers {
+		for _, filter := range filters {
+			if !filter(invoker) {
+				continue OUTER
```
Comment (AlexStocks) [inline cluster/router/tag/tag_router.go:275]: break OUTER？

File: cluster/router/tag/tag_router.go (hunk 13 above)
Comment (AlexStocks) [reply cluster/router/tag/tag_router.go:332]: we should not use Chinese comment in our codes.

File: cluster/router/tag/tag_router.go [hunk 16]
```diff
@@ -100,3 +263,163 @@ func isForceUseTag(url *common.URL, invocation protocol.Invocation) bool {
...
+	}
+
+	// short name condition
+	if !ipPatternContains(pattern) {
+		return pattern == host
+	}
+
+	ipAddress := strings.Split(host, splitCharacter)
+	for i := 0; i < len(mask); i++ {
+		if "*" == mask[i] || mask[i] == ipAddress[i] {
```
Comment (AlexStocks) [inline cluster/router/tag/tag_router.go:356]: u should add comment for every if/else branch to make it easy to understand.

File: cluster/router/tag/tag_router.go [hunk 17]
```diff
@@ -100,3 +263,163 @@ func isForceUseTag(url *common.URL, invocation protocol.Invocation) bool {
...
+		if len(mask) != 4 {
+			return errors.New("The host is ipv4, but the pattern is not ipv4 pattern : " + pattern)
+		}
+	}
+	return nil
+}
+
+func getPatternHostAndPort(pattern string, isIpv4 bool) []string {
+	result := make([]string, 2)
+	if strings.HasPrefix(pattern, "[") && strings.Contains(pattern, "]:") {
```
Comment (AlexStocks) [inline cluster/router/tag/tag_router.go:401]: add some comments for every if/else branch, pls.

File: cluster/router/tag/router_rule.go [hunk 18]
```diff
@@ -34,5 +52,58 @@ func getRule(rawRule string) (*RouterRule, error) {
 		return r, err
 	}
 	r.RawRule = rawRule
+	r.init()
 	return r, nil
 }
+
+func (t *RouterRule) init() {
```
Comment (hxmhlt) [inline cluster/router/tag/router_rule.go:59]: Why create a new func named init? If you create the func ,pls add the comment for it.

File: cluster/router/tag/router_rule.go [hunk 19]
```diff
@@ -34,5 +52,58 @@ func getRule(rawRule string) (*RouterRule, error) {
...
+	for _, tag := range t.Tags {
+		for _, address := range tag.Addresses {
+			t.addressToTagNames[address] = append(t.addressToTagNames[address], tag.Name)
+		}
+		t.tagNameToAddresses[tag.Name] = tag.Addresses
+	}
+}
+
+func (t *RouterRule) getAddresses() []string {
+	var result = make([]string, 0, 8*len(t.Tags))
```
Comment (hxmhlt) [inline cluster/router/tag/router_rule.go:71]: Pls add func comment for RouterRule

File: cluster/router/tag/router_rule.go [hunk 20]
```diff
@@ -34,5 +52,58 @@ func getRule(rawRule string) (*RouterRule, error) {
 		return r, err
 	}
 	r.RawRule = rawRule
+	r.init()
 	return r, nil
 }
+
+func (t *RouterRule) init() {
+	t.addressToTagNames = make(map[string][]string, 8)
+	t.tagNameToAddresses = make(map[string][]string, 8)
```
Comment (zouyx) [inline cluster/router/tag/router_rule.go:61]:
```suggestion
         l:=len(t.Tags)
	t.addressToTagNames = make(map[string][]string, l*2)
	t.tagNameToAddresses = make(map[string][]string, l)
```

File: cluster/router/tag/router_rule.go (hunk 19 above)
Comment (zouyx) [inline cluster/router/tag/router_rule.go:71]:
```suggestion
	var result = make([]string, 0, 2*len(t.Tags))
```
i think * 2 enough

File: cluster/router/tag/router_rule.go [hunk 21]
```diff
@@ -34,5 +52,58 @@ func getRule(rawRule string) (*RouterRule, error) {
...
+	return result
+}
+
+func (t *RouterRule) hasTag(tag string) bool {
+	for _, t := range t.Tags {
+		if tag == t.Name {
+			return true
+		}
+	}
+	return false
```
Comment (zouyx) [inline cluster/router/tag/router_rule.go:92]:
```suggestion
        return len(tagNameToAddresses[tag])>0
```

Why dont you use `tagNameToAddresses    map[string][]string`?

File: cluster/router/tag/tag_router.go [hunk 22]
```diff
@@ -55,6 +64,15 @@ func (c *tagRouter) isEnabled() bool {
 	return c.enabled
 }
 
+func (c *tagRouter) SetApplication(app string) {
+	c.application = app
+}
+
+func (c *tagRouter) tagRouterRuleCopy() RouterRule {
+	routerRule := *c.tagRouterRule
+	return routerRule
```
Comment (zouyx) [inline cluster/router/tag/tag_router.go:73]:
```suggestion
	return *c.tagRouterRule
```

enough? and i think no need to extract a new function for single scene

File: cluster/router/tag/tag_router.go [hunk 23]
```diff
@@ -63,7 +81,152 @@ func (c *tagRouter) Route(invokers []protocol.Invoker, url *common.URL, invocati
...
+				if len(addresses) == 0 || !checkAddressMatch(tagRouterRuleCopy.getAddresses(), url.Ip, url.Port) {
+					return true
+				}
+				return false
+			}
+			filterTagIsEmpty := func(invoker protocol.Invoker) bool {
+				if invoker.GetUrl().GetParam(constant.Tagkey, "") == "" {
+					return true
+				}
+				return false
```
Comment (zouyx) [inline cluster/router/tag/tag_router.go:141]:
```suggestion
				return invoker.GetUrl().GetParam(constant.Tagkey, "") == "" 
```

File: cluster/router/tag/tag_router.go [hunk 24]
```diff
@@ -63,7 +81,152 @@ func (c *tagRouter) Route(invokers []protocol.Invoker, url *common.URL, invocati
...
+	} else {
+		// return all addresses in dynamic tag group.
+		addresses = tagRouterRuleCopy.getAddresses()
+		if len(addresses) > 0 {
+			filterAddressNotMatches := func(invoker protocol.Invoker) bool {
+				url := invoker.GetUrl()
+				if len(addresses) == 0 || !checkAddressMatch(addresses, url.Ip, url.Port) {
+					return true
+				}
+				return false
```
Comment (zouyx) [inline cluster/router/tag/tag_router.go:154]:
```suggestion
				return len(addresses) == 0 || !checkAddressMatch(addresses, url.Ip, url.Port)
```

File: cluster/router/tag/tag_router.go [hunk 25]
```diff
@@ -63,7 +81,152 @@ func (c *tagRouter) Route(invokers []protocol.Invoker, url *common.URL, invocati
...
+	if len(tag) > 0 {
+		addresses, _ = tagRouterRuleCopy.getTagNameToAddresses()[tag]
+		// filter by dynamic tag group first
+		if len(addresses) > 0 {
+			filterAddressMatches := func(invoker protocol.Invoker) bool {
+				url := invoker.GetUrl()
+				if len(addresses) > 0 && checkAddressMatch(addresses, url.Ip, url.Port) {
+					return true
+				}
+				return false
```
Comment (zouyx) [inline cluster/router/tag/tag_router.go:107]:
```suggestion
				return len(addresses) > 0 && checkAddressMatch(addresses, url.Ip, url.Port) 
```

File: cluster/router/tag/tag_router.go [hunk 26]
```diff
@@ -63,7 +81,152 @@ func (c *tagRouter) Route(invokers []protocol.Invoker, url *common.URL, invocati
...
+				return result
+			}
+		} else {
+			// dynamic tag group doesn't have any item about the requested app OR it's null after filtered by
+			// dynamic tag group but force=false. check static tag
+			filter := func(invoker protocol.Invoker) bool {
+				if invoker.GetUrl().GetParam(constant.Tagkey, "") == tag {
+					return true
+				}
+				return false
```
Comment (zouyx) [inline cluster/router/tag/tag_router.go:120]:
```suggestion
				return invoker.GetUrl().GetParam(constant.Tagkey, "") == tag 
```

File: cluster/router/tag/tag_router.go [hunk 27]
```diff
@@ -63,7 +81,152 @@ func (c *tagRouter) Route(invokers []protocol.Invoker, url *common.URL, invocati
...
+		if len(result) > 0 || isForceUseTag(url, invocation) {
+			return result
+		} else {
+			// FAILOVER: return all Providers without any tags.
+			filterAddressNotMatches := func(invoker protocol.Invoker) bool {
+				url := invoker.GetUrl()
+				if len(addresses) == 0 || !checkAddressMatch(tagRouterRuleCopy.getAddresses(), url.Ip, url.Port) {
+					return true
+				}
+				return false
```
Comment (zouyx) [inline cluster/router/tag/tag_router.go:135]:
```suggestion
				return len(addresses) == 0 || !checkAddressMatch(tagRouterRuleCopy.getAddresses(), url.Ip, url.Port)
```

File: cluster/router/tag/tag_router.go [hunk 28]
```diff
@@ -63,7 +81,152 @@ func (c *tagRouter) Route(invokers []protocol.Invoker, url *common.URL, invocati
...
+		return filterInvoker(result, filter)
+	}
+}
+
+func (c *tagRouter) Process(event *config_center.ConfigChangeEvent) {
+	logger.Infof("Notification of tag rule, change type is:[%s] , raw rule is:[%v]", event.ConfigType, event.Value)
+	if remoting.EventTypeDel == event.ConfigType {
+		c.tagRouterRule = nil
+		return
+	} else {
```
Comment (zouyx) [inline cluster/router/tag/tag_router.go:177]: delete else

File: cluster/router/tag/tag_router.go [hunk 29]
```diff
@@ -100,3 +263,163 @@ func isForceUseTag(url *common.URL, invocation protocol.Invocation) bool {
...
+}
+
+func matchIp(pattern, host, port string) bool {
+	// if the pattern is subnet format, it will not be allowed to config port param in pattern.
+	if strings.Contains(pattern, "/") {
+		_, subnet, _ := net.ParseCIDR(pattern)
+		if subnet != nil && subnet.Contains(net.ParseIP(host)) {
+			return true
+		}
+		return false
```
Comment (zouyx) [inline cluster/router/tag/tag_router.go:303]:
```suggestion
		return subnet != nil && subnet.Contains(net.ParseIP(host))
```

File: cluster/router/tag/tag_router_test.go [hunk 30]
```diff
@@ -19,32 +19,59 @@ package tag
...
 	"testing"
+	"time"
 )
 
 import (
+	"github.com/dubbogo/go-zookeeper/zk"
```
Comment (zouyx) [inline cluster/router/tag/tag_router_test.go:29]: split package

**Pull request merged**